crypto/ec/ecp_s390x_nistp.c

```diff
@@ -0,0 +1,394 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <openssl/err.h>
+#include <openssl/rand.h>
+#include "ec_lcl.h"
+#include "s390x_arch.h"
+
+/* Size of parameter blocks */
+#define S390X_SIZE_PARAM		4096
+
+/* Size of fields in parameter blocks */
+#define S390X_SIZE_P256			32
+#define S390X_SIZE_P384			48
+#define S390X_SIZE_P521			80
+
+/* Offsets of fields in PCC parameter blocks */
+#define S390X_OFF_RES_X(n)		(0 * n)
+#define S390X_OFF_RES_Y(n)		(1 * n)
+#define S390X_OFF_SRC_X(n)		(2 * n)
+#define S390X_OFF_SRC_Y(n)		(3 * n)
+#define S390X_OFF_SCALAR(n)		(4 * n)
+
+/* Offsets of fields in KDSA parameter blocks */
+#define S390X_OFF_R(n)			(0 * n)
+#define S390X_OFF_S(n)			(1 * n)
+#define S390X_OFF_H(n)			(2 * n)
+#define S390X_OFF_K(n)			(3 * n)
+#define S390X_OFF_X(n)			(3 * n)
+#define S390X_OFF_RN(n)			(4 * n)
+#define S390X_OFF_Y(n)			(4 * n)
+
+static int ec_GFp_s390x_nistp_mul(const EC_GROUP *group, EC_POINT *r,
+                                  const BIGNUM *scalar,
+                                  size_t num, const EC_POINT *points[],
+                                  const BIGNUM *scalars[],
+                                  BN_CTX *ctx, unsigned int fc, int len)
+{
+    unsigned char param[S390X_SIZE_PARAM];
+    BIGNUM *x, *y;
+    const EC_POINT *point_ptr = NULL;
+    const BIGNUM *scalar_ptr = NULL;
+    BN_CTX *new_ctx = NULL;
+    int rc = -1;
+
+    if (ctx == NULL) {
+        ctx = new_ctx = BN_CTX_new();
+        if (ctx == NULL)
+            return 0;
+    }
+
+    BN_CTX_start(ctx);
+
+    x = BN_CTX_get(ctx);
+    y = BN_CTX_get(ctx);
+    if (x == NULL || y == NULL) {
+        rc = 0;
+        goto ret;
+    }
+
+    /*
+     * Use PCC for EC keygen and ECDH key derivation:
+     * scalar * generator and scalar * peer public key,
+     * scalar in [0,order).
+     */
+    if ((scalar != NULL && num == 0 && BN_is_negative(scalar) == 0)
+        || (scalar == NULL && num == 1 && BN_is_negative(scalars[0]) == 0)) {
+
+        if (num == 0) {
+            point_ptr = EC_GROUP_get0_generator(group);
+            scalar_ptr = scalar;
+        } else {
+            point_ptr = points[0];
+            scalar_ptr = scalars[0];
+        }
+
+        if (EC_POINT_is_at_infinity(group, point_ptr) == 1
+            || BN_is_zero(scalar_ptr)) {
```

Comment:
General ECC question ..

ECDH and ECDSA keys (ie the cases where scalar is secret) are in [1,order), so i thought i could just special-case scalar==0 here. However, the comment at EC_METHOD's mul function (ec_lcl.h) says:
```
     * For a fixed point multiplication (scalar != NULL, num == 0)
     * or a variable point multiplication (scalar == NULL, num == 1),
     * mul() must use a constant time algorithm: in both cases callers                         
     * should provide an input scalar (either scalar or scalars[0])
     * in the range [0, ec_group_order); ...
```
So what can go wrong sca-wise, if scalar==0 case can be distinguished ?

Reply:
@romen ?